core/ptr: Add simulate_realloc()

mrkajetanp opened the pull request:
Add a `simulate_realloc()` helper function to core/ptr to simulate
reallocating memory from a pointer to some arbitrary address.
The function is intended to be used with architecture features such as
AArch64 Top-Byte Ignore where two different 64-bit addresses can point
to the same chunk of memory due to some bits being ignored.

To make the function accurately simulate an allocator, its return value
needs to be annotated with `noalias` in LLVM the same way as it is done
by actual allocator functions. To make that possible, add a new rustc
built-in attribute `rustc_simulate_allocator` that does the annotating.

Accompanying implementation for [RFC 3700](https://github.com/rust-lang/rfcs/pull/3700), currently posted on [Rust Internals](https://internals.rust-lang.org/t/pre-rfc-core-simulate-realloc/21745).

r? @RalfJung 

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11056347577/job/30717693830) [(plain)](https://github.com/rust-lang/rust/commit/c4c2b4e8da197f9bb91d65b3849b981d6c371b10/checks/30717693830/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.803 Building wheels for collected packages: reuse
#16 2.804   Building wheel for reuse (pyproject.toml): started
#16 3.047   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.048   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.049   Stored in directory: /tmp/pip-ephem-wheel-cache-507us52t/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.051 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.444 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.444 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.5s
---
   Compiling std v0.0.0 (/checkout/library/std)
error: this arithmetic operation will overflow
    --> core/src/ptr/const_ptr.rs:1544:9
     |
1544 |         (self.addr() >> 56) as u8
     |         ^^^^^^^^^^^^^^^^^^^ attempt to shift right by `56_i32`, which would overflow
     |
     = note: `#[deny(arithmetic_overflow)]` on by default
error: this arithmetic operation will overflow
    --> core/src/ptr/const_ptr.rs:1563:60
     |
     |
1563 |         self.map_addr(|addr| (addr & 0x00ffffffffffffff) | (tag as usize) << 56)
     |                                                            ^^^^^^^^^^^^^^^^^^^^ attempt to shift left by `56_i32`, which would overflow
error: this arithmetic operation will overflow
    --> core/src/ptr/mut_ptr.rs:1794:9
     |
1794 |         (self.addr() >> 56) as u8
1794 |         (self.addr() >> 56) as u8
     |         ^^^^^^^^^^^^^^^^^^^ attempt to shift right by `56_i32`, which would overflow
error: this arithmetic operation will overflow
    --> core/src/ptr/mut_ptr.rs:1813:60
     |
     |
1813 |         self.map_addr(|addr| (addr & 0x00ffffffffffffff) | (tag as usize) << 56)
     |                                                            ^^^^^^^^^^^^^^^^^^^^ attempt to shift left by `56_i32`, which would overflow
error: literal out of range for `usize`
    --> core/src/ptr/const_ptr.rs:1563:38
     |
     |
1563 |         self.map_addr(|addr| (addr & 0x00ffffffffffffff) | (tag as usize) << 56)
     |
     |
     = note: the literal `0x00ffffffffffffff` (decimal `72057594037927935`) does not fit into the type `usize` and will become `4294967295usize`
     = note: `#[deny(overflowing_literals)]` on by default
error: literal out of range for `usize`
    --> core/src/ptr/mut_ptr.rs:1813:38
     |
     |
1813 |         self.map_addr(|addr| (addr & 0x00ffffffffffffff) | (tag as usize) << 56)
     |
     |
     = note: the literal `0x00ffffffffffffff` (decimal `72057594037927935`) does not fit into the type `usize` and will become `4294967295usize`
error: could not compile `core` (lib) due to 6 previous errors
Build completed unsuccessfully in 0:01:03
  local time: Thu Sep 26 16:33:59 UTC 2024
  network time: Thu, 26 Sep 2024 16:33:59 GMT
```

</details>

mrkajetanp commented:
This is just a reference PR for the RFC and the implementation is aarch64-specific hence the test failures. The question I hope the RFC process answers is whether the best approach would be to have a generic method like this with cfgs to choose between variants for different architectures hidden inside, or whether it should just be completely architecture-specific and put away in `core_arch` so that the users themselves have to opt-in to each architecture and/or tagging scheme they want to support.

mrkajetanp committed:
core/ptr: Add simulate_realloc()

Add a `simulate_realloc()` helper function to core/ptr to simulate
reallocating memory from a pointer to some arbitrary address.
The function is intended to be used with architecture features such as
AArch64 Top-Byte Ignore where two different 64-bit addresses can point
to the same chunk of memory due to some bits being ignored.

To make the function accurately simulate an allocator, its return value
needs to be annotated with `noalias` in LLVM the same way as it is done
by actual allocator functions. To make that possible, add a new rustc
built-in attribute `rustc_simulate_allocator` that does the annotating.

programmerjake commented on library/core/src/ptr/mod.rs:
```diff
@@ -2441,3 +2442,33 @@ pub macro addr_of($place:expr) {
+    // SAFETY: This does not do anything
+    unsafe {
+        asm!("/* simulate realloc from {original} to {ptr} */",
+         original = in(reg) original, ptr = inout(reg) ptr);
```
this looks like non-standard formatting to me, did this get run through rustfmt? does rustfmt actually format this or just leave whatever you do in place? I would have written it:
```rust
        asm!(
            "/* simulate realloc from {original} to {ptr} */",
            original = in(reg) original,
            ptr = inout(reg) ptr,
        );
```

bors commented:
:umbrella: The latest upstream changes (presumably #135370) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
